dbos/system_database.go

```diff
@@ -756,6 +774,87 @@ func (s *sysDB) cancelWorkflow(ctx context.Context, workflowID string) error {
 	return nil
 }
 
+func (s *sysDB) cancelAllBefore(ctx context.Context, cutoffTime time.Time) error {
+	// List all workflows in PENDING or ENQUEUED state ending at cutoffTime
+	listInput := listWorkflowsDBInput{
+		endTime: cutoffTime,
+		status:  []WorkflowStatusType{WorkflowStatusPending, WorkflowStatusEnqueued},
+	}
+
+	workflows, err := s.listWorkflows(ctx, listInput)
+	if err != nil {
+		return fmt.Errorf("failed to list workflows for cancellation: %w", err)
+	}
+
+	// Cancel each workflow
+	for _, workflow := range workflows {
+		if err := s.cancelWorkflow(ctx, workflow.ID); err != nil {
+			s.logger.Error("Failed to cancel workflow during cancelAllBefore", "workflowID", workflow.ID, "error", err)
+			// Continue with other workflows even if one fails
+			// If desired we could funnel the errors back the caller (conductor, admin server)
+		}
+	}
+
+	return nil
+}
+
+func (s *sysDB) garbageCollectWorkflows(ctx context.Context, input garbageCollectWorkflowsInput) error {
+	// Validate input parameters
+	if input.rowsThreshold != nil && *input.rowsThreshold <= 0 {
+		return fmt.Errorf("rowsThreshold must be greater than 0, got %d", *input.rowsThreshold)
+	}
```

Comment:
Technically this could be enforced by using a unsigned integer, but I trust it makes for a better error message.